xapian-applications/omega/query.cc

```diff
@@ -651,24 +652,28 @@ html_strip(const string &str)
     return res;
 }
 
-// FIXME split list into hash or map and use that rather than linear lookup?
-static int word_in_list(const string& word, const string& list)
-{
-    string::size_type split = 0, split2;
+static unordered_map<string, int> word_to_occurrence;
+
+static void build_word_map(const string& list) {
+    string word;
     int count = 0;
-    while ((split2 = list.find('\t', split)) != string::npos) {
-	if (word.size() == split2 - split) {
-	    if (memcmp(word.data(), list.data() + split, word.size()) == 0)
-		return count;
-	}
-	split = split2 + 1;
-	++count;
+    for (size_t i = 0; i <= list.length(); ++i) {
+	if (list[i] == '\t' || list[i] == '\0') {
+	    if (word_to_occurrence.find(word) == word_to_occurrence.end()) {
```

Comment:
I wasn't actually aware there was a `count()` method on `std::map`, etc!

I agree `map.count(key)` is more concise source code than `map.find(key) != map.end()`.  For the `std::map` implementation in GCC7, `count()` actually just inlines `map.find(key) == map.end() ? 0 : 1` so the two should end up producing equivalently good code.

However, while it's true that you aren't actually wanting to find it, you also really aren't wanting to count the number of occurrences - you just want to know if there are any, not how many there are.  And for `std::multimap` that distinction matters, as `count()` has to do more work that the find approach (because there `count()` calls `equal_range()` and then calculates the distance between the two iterators returned, and `std::multimap::iterator` is bi-directional, so that makes `count()` O(n) where n is the number of items with the requested key).

Exactly what the implementation of each is for `std::unordered_map` is less clear to me, and compiling a simple test function and looking at the code only really told me that the two approaches don't result in the same generated code.  But at least in the "multi" case it seems to me that `count()` will have to exhaustively check the bucket which the requested key is in, which `find()` can stop once it finds an instance of the key.

So for consistency I'd probably suggest we stick with `map.find(key) != map.end()` so that there's a single style rule for the idiom to use for this, independent of whether the container is a "multi" one or not.